src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -138,6 +93,145 @@ void GcsNodeManager::ScheduleTick() {
   });
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+GcsNodeManager::GcsNodeManager(boost::asio::io_service &io_service,
+                               gcs::NodeInfoAccessor &node_info_accessor,
+                               gcs::ErrorInfoAccessor &error_info_accessor)
+    : node_info_accessor_(node_info_accessor),
+      error_info_accessor_(error_info_accessor),
+      node_failure_detector_(new NodeFailureDetector(io_service, node_info_accessor)) {
+  // TODO(Shanly): Load node info list from storage synchronously.
+  const auto lookup_callback =
+      [this](Status status, const std::vector<rpc::GcsNodeInfo> &node_info_list) {
+        for (const auto &node_info : node_info_list) {
+          if (node_info.state() != rpc::GcsNodeInfo::DEAD) {
+            AddNode(std::make_shared<rpc::GcsNodeInfo>(node_info));
+          }
+        }
+      };
+  RAY_CHECK_OK(node_info_accessor_.AsyncGetAll(lookup_callback));
+  // TODO(Shanly): Load cluster resources from storage synchronously.
+  node_failure_detector_->AddNodeDeadListener([this](const ClientID &node_id) {
+    if (auto node = RemoveNode(node_id, /* is_intended = */ false)) {
+      RAY_CHECK_OK(node_info_accessor_.AsyncUnregister(node_id, nullptr));
+    }
+  });
+}
+
+void GcsNodeManager::HandleRegisterNode(const rpc::RegisterNodeRequest &request,
+                                        rpc::RegisterNodeReply *reply,
+                                        rpc::SendReplyCallback send_reply_callback) {
+  ClientID node_id = ClientID::FromBinary(request.node_info().node_id());
+  RAY_LOG(INFO) << "Registering node info, node id = " << node_id;
+  auto on_done = [this, node_id, request, reply, send_reply_callback](Status status) {
+    RAY_CHECK_OK(status);
+    RAY_LOG(INFO) << "Finished registering node info, node id = " << node_id;
+    AddNode(std::make_shared<rpc::GcsNodeInfo>(request.node_info()));
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
```

Comment:
is it possible to just call `send_reply_callback` instead, like what's done in other components?  
`GCS_RPC_SEND_REPLY` seems like something that would only be used in rpc related modules.

Reply:
> is it possible to just call `send_reply_callback` instead, like what's done in other components?
> `GCS_RPC_SEND_REPLY` seems like something that would only be used in rpc related modules.

Now, GcsNodeManager takes over the RPC messages about nodes.